File: 256_paint_house.java
class Solution {
    public int minCost(int[][] costs) {
        int[][] dp = new int[costs.length + 1][3];
        int houses = costs.length;
        for (int i = 1; i <= houses; i++) {
            dp[i][0] = Math.min(dp[i - 1][1], dp[i - 1][2]) + costs[i - 1][0];
            dp[i][1] = Math.min(dp[i - 1][0], dp[i - 1][2]) + costs[i - 1][1];
            dp[i][2] = Math.min(dp[i - 1][1], dp[i - 1][0]) + costs[i - 1][2];
        }
        return Math.min(dp[costs.length][0], Math.min(dp[costs.length][1], dp[costs.length][2]));
    }
}

File: python/256_paint_house.py
class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        N = len(costs)
        if N == 0:
            return 0

        now = [0] * 3
        prev = costs[0][:]

        for i in range(N - 1):
            now[0] = min(prev[1], prev[2]) + costs[i + 1][0]
            now[1] = min(prev[0], prev[2]) + costs[i + 1][1]
            now[2] = min(prev[0], prev[1]) + costs[i + 1][2]
            prev[:] = now[:]
        return min(prev)

